astropy/uncertainty/core.py

```diff
@@ -100,9 +100,9 @@ def __new__(cls, samples):
             interface["strides"] = samples.strides[:-1]
             structured = np.asarray(DummyArray(interface, base=samples))
         # Set our new structured dtype.
-        structured.dtype = new_dtype
+        structured = structured.view(new_dtype)
```

Comment:
This is probably not quite it for `uncertainty`, correct? If so, I'd prefer to do `uncertainty` separately -- maybe for that one it is worth using the private methods.